### PR TITLE
Zoom Toolbar

### PR DESCRIPTION
**Pull request by alangerak**
Added a toolbar with 2 buttons and a slider. Those can be used to zoom in/out on the graph.

### CODE REVIEW

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/CustomLabelSlider.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+            public String toString(final Double number) {
+                return String.format("%1$.0f", function.apply(number));
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public Double fromString(final String string) {
+                return null;
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/CustomLabelSlider.java]: Is this desired?

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 2]
```diff
@@ -127,15 +131,51 @@
...
+    private void initZoomToolBar() {
+        ZoomToolbar toolbar = new ZoomToolbar(MAXZOOM);
+        wrapper.setRight(toolbar.getToolBar());
+
+        toolbar.getZoomlevel().addListener((observeVal, oldVal, newVal) -> {
+
+            int diffLevel = oldVal.intValue() - newVal.intValue();
+            zoomLevel = Math.abs(newVal.intValue());
+            if (diffLevel < 0) {
+                zoom(Math.pow(ZOOMOUTFACTOR, diffLevel * -1));
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java]: duplicate code, could use a method

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java [hunk 3]
```diff
@@ -291,12 +316,12 @@ private void repaint() {
             }
             view = new TileView(this);
 
-            wrapper.hvalueProperty().addListener(
+            scrollPane.hvalueProperty().addListener(
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java:316]: Useless curly brackets

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    public ZoomToolbar(final int zoom) {
+        maxzoom = zoom;
+        zoomLevel = new SimpleIntegerProperty();
+        // 5 is temporary, until graph can start at the highest zoom level
+        zoomLevel.set(5);
+        Slider slider = createSlider();
+        slider.getStyleClass().add("slider");
+        slider.valueProperty().addListener(
+                (ChangeListener<Number>) (obserVal, oldVal, newVal) -> {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: I think `(ChangeListener<Number>)` is redundant

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                    if (oldVal.intValue() > newVal.intValue()
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: Combine these if statements with `oldVal.intValue() != newVal.intValue()`

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                            && newVal.intValue() <= maxzoom) {
+                        zoomLevel.set(newVal.intValue());
+                    }
+                    if (oldVal.intValue() < newVal.intValue()) {
+                        zoomLevel.set(newVal.intValue());
+                    }
+                });
+
+        Button plus = new Button("+");
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: Duplicate code, use a method.

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    private final int maxzoom;
+
+    /**
+     * Create a new Toolbar.
+     *
+     * @param zoom
+     *            the maximal value of the zoom
+     */
+    public ZoomToolbar(final int zoom) {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: You could (and probably should) extend `ToolBar`

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        slider.setValue(zoomLevel.intValue());
+        slider.setMax(maxzoom);
+        slider.setMajorTickUnit(1);
+        slider.setMinorTickCount(0);
+        slider.setShowTickMarks(true);
+        slider.setShowTickLabels(true);
+        slider.setSnapToTicks(true);
+        slider.setMinHeight(Slider.USE_PREF_SIZE);
+
+        slider.valueChangingProperty().addListener(observable -> {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: Useless curly brackets

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/view/VertexView.java [hunk 9]
```diff
@@ -123,16 +129,20 @@ protected final void layoutChildren() {
         double height = rectangle.getHeight();
 
         double fontWidth = text.getLayoutBounds().getWidth();
-
         text.setFont(Font.font(FONTNAME, (HORIZONTALSCALE) * width / fontWidth));
+        text.setLayoutX(width / 2 - text.getLayoutBounds().getWidth() / 2);
+        text.setLayoutY(height / 2);
+
+        if ((HORIZONTALSCALE * width / fontWidth) < MINTEXTSIZE) {
```
Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/view/VertexView.java]: Can be refactored into a one-liner: `text.setVisible(HORIZONTALSCALE * width / fontWidth >= MINTEXTSIZE)`.

File: lifetiles-launcher/src/main/resources/styles/ToolBarStyle.css [hunk 10]
```diff
@@ -0,0 +1,19 @@
+/* Stylesheet for the Toolbar */
+
+.plusButton{
+    -fx-background-color: rgb(100.0,150.0,255.0);
```
Comment (jorenham) [inline lifetiles-launcher/src/main/resources/styles/ToolBarStyle.css]: Duplicate color, you can use 

``` css
* {
   -fx-defaultcolor: rgb(100.0,150.0,255.0);
}
.plusButton{
    -fx-background-color: -fx-defaultcolor;
}
```

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java (hunk 7 above)
Comment (alangerak) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: This class is not an extension of the ToolBar class but rather an implementation of it with added buttons and a custom slider. It does not add new functionalities to ToolBar.

**Comment by albert-smit** [comment]
looks good, but where is coveralls?

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java (hunk 3 above)
Comment (alangerak) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java:316]: tried to remove then, but the formatter aligns the code in a very strange way without the brackets.

Comment (jorenham) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java:316]: that's not a good reason to not remove redundant code, you could always format it yourself

File: lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java (hunk 6 above)
Comment (alangerak) [inline lifetiles-graph/src/main/java/nl/tudelft/lifetiles/graph/controller/ZoomToolbar.java]: What duplicate code?

**Pull request merged**